### PR TITLE
Trip tracking

### PR DESCRIPTION
**Pull request by nickpatrick**
https://app.clubhouse.io/radarlabs/story/2494/trip-tracking-ios

```swift
let options = RadarTripOptions(externalId: "abc")
options.metadata = ["foo": "bar"]
options.destinationGeofenceTag = "store"
options.destinationGeofenceExternalId = "123"
options.mode = .car
Radar.startTrip(options: options)
        
Radar.stopTrip()
```

### CODE REVIEW

File: RadarSDK/RadarTripOptions.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ @see https://radar.io/documentation/sdk#ios-trips
+ */
+@interface RadarTripOptions : NSObject
+
+- (instancetype)initWithExternalId:(NSString *_Nonnull)externalId;
+
+/**
+ A stable unique ID for the trip.
+ */
+@property (nullable, nonatomic, copy) NSString *externalId;
```
Comment (pingxia0823) [inline RadarSDK/RadarTripOptions.h:26]: nonnull?

File: RadarSDK/RadarTripOptions.m [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+- (NSDictionary *)dictionaryValue {
+    NSMutableDictionary *dict = [NSMutableDictionary new];
+    dict[kExternalId] = self.externalId;
+    dict[kMetadata] = self.metadata;
+    dict[kDestinationGeofenceTag] = self.destinationGeofenceTag;
+    dict[kDestinationGeofenceExternalId] = self.destinationGeofenceExternalId;
+    dict[kMode] = [Radar stringForMode:self.mode];
+    return dict;
```
Comment (pingxia0823) [inline RadarSDK/RadarTripOptions.m:54]: we may need to have nil guards here as several properties could be nil

File: RadarSDK/RadarTripOptions.m [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    options.destinationGeofenceExternalId = dict[kDestinationGeofenceExternalId];
+    NSString *modeStr = dict[kMode];
+    if ([modeStr isEqualToString:@"foot"]) {
+        options.mode = RadarRouteModeFoot;
+    } else if ([modeStr isEqualToString:@"bike"]) {
+        options.mode = RadarRouteModeBike;
+    } else {
+        options.mode = RadarRouteModeCar;
+    }
+    return options;
```
Comment (pingxia0823) [inline RadarSDK/RadarTripOptions.m:44]: nil checking may be needed for @"mode" and "externalId"

File: RadarSDK/RadarAPIClient.m [hunk 4]
```diff
@@ -155,6 +156,16 @@ - (void)trackWithLocation:(CLLocation *_Nonnull)location
...
+    RadarTripOptions *tripOptions = [RadarSettings tripOptions];
+    if (tripOptions) {
+        params[@"tripExternalId"] = tripOptions.externalId;
+        params[@"tripMetadata"] = tripOptions.metadata;
+        if (tripOptions.destinationGeofenceTag && tripOptions.destinationGeofenceExternalId) {
+            params[@"tripDestinationGeofenceTag"] = tripOptions.destinationGeofenceTag;
+            params[@"tripDestinationGeofenceExternalId"] = tripOptions.destinationGeofenceExternalId;
+            params[@"tripMode"] = [Radar stringForMode:tripOptions.mode];
+        }
+    }
```
Comment (pingxia0823) [inline RadarSDK/RadarAPIClient.m:168]: wonder if we shall put them under 

```
params[@"tripOptions"] = @{
@"externalId": xxx ///
@"mode": yyy ///
...
}
```

File: RadarSDK/RadarTripOptions.h (hunk 1 above)
Comment (nickpatrick) [reply RadarSDK/RadarTripOptions.h:26]: Good catch

File: RadarSDK/RadarAPIClient.m (hunk 4 above)
Comment (nickpatrick) [reply RadarSDK/RadarAPIClient.m:168]: Open to this, let's discuss tradeoffs when we build the server piece

File: RadarSDK/RadarTripOptions.m (hunk 3 above)
Comment (nickpatrick) [reply RadarSDK/RadarTripOptions.m:44]: I'll add `nil` checks for `externalId` and `modeStr` to be extra safe

File: RadarSDK/Radar.h [hunk 5]
```diff
@@ -567,6 +585,15 @@ typedef void (^_Nonnull RadarRouteCompletionHandler)(RadarStatus status, RadarRo
...
 + (NSString *)stringForSource:(RadarLocationSource)source NS_SWIFT_NAME(stringForSource(_:));
 
+/**
+ Returns a display string for a travel mode value.
+
+ @param mode A travel mode value.
+
+ @return A display string for the travel mode value.
+ */
++ (NSString *)stringForMode:(RadarRouteMode)mode NS_SWIFT_NAME(stringForMode(_:));
```
Comment (jsani-radar) [inline RadarSDK/Radar.h:595]: is this better suited in the `RadarRoute` class?

File: RadarSDK/RadarTripOptions.m (hunk 2 above)
Comment (jsani-radar) [reply RadarSDK/RadarTripOptions.m:54]: I think if `tripOptionsFromDictionary` returns `nil` without `externalId` / `mode`, then this function should be fine with just a check at the top for `self == nil`

File: RadarSDK/RadarTripOptions.m [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        return YES;
+    }
+
+    if (![object isKindOfClass:[RadarTripOptions class]]) {
+        return NO;
+    }
+
+    RadarTripOptions *options = (RadarTripOptions *)object;
+
+    return [self.externalId isEqualToString:options.externalId] &&
```
Comment (jsani-radar) [inline RadarSDK/RadarTripOptions.m:72]: wonder if there's a way to autogenerate these, could be good to add to `RadarCollectionAdditions` /cc @pingxia0823

File: RadarSDK/RadarAPIClient.m (hunk 4 above)
Comment (pingxia0823) [reply RadarSDK/RadarAPIClient.m:168]: https://github.com/radarlabs/server/blob/master/app.ts#L1002
https://github.com/radarlabs/server/blob/master/server/services/UserUpsertService.ts#L335
Server expects a `tripOptions` dictionary wrapping all trip related fields. The main benefit of this is not to bloat the userUpsertService interface, and make it easier for validity checks

```
params[@"tripOptions"] = @{
@"externalId": ... // required
@"mode": ... // required
@"destinationGeofenceTag": ...
@"destinationGeofenceExternalId": ...
@"metadata": ...
}
```

Comment (nickpatrick) [reply RadarSDK/RadarAPIClient.m:168]: Awesome, I will update this

**Pull request merged**